src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/MessagePackHubProtocol.java

```diff
@@ -0,0 +1,572 @@
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.msgpack.core.MessageBufferPacker;
+import org.msgpack.core.MessageFormat;
+import org.msgpack.core.MessagePack;
+import org.msgpack.core.MessagePackException;
+import org.msgpack.core.MessagePacker;
+import org.msgpack.core.MessageUnpacker;
+import org.msgpack.value.ValueType;
+
+class MessagePackHubProtocol implements HubProtocol {
+    
+    private static final int ERROR_RESULT = 1;
+    private static final int VOID_RESULT = 2;
+    private static final int NON_VOID_RESULT = 3;
+
+    @Override
+    public String getName() {
+        return "messagepack";
+    }
+
+    @Override
+    public int getVersion() {
+        return 1;
+    }
+
+    @Override
+    public TransferFormat getTransferFormat() {
+        return TransferFormat.BINARY;
+    }
+
+    @Override
+    public HubMessage[] parseMessages(String payload, InvocationBinder binder) {
+        if (payload.length() == 0) {
+            return new HubMessage[]{};
+        }
+
+        List<HubMessage> hubMessages = new ArrayList<>();
+        
+        try {
+            MessageUnpacker unpacker = MessagePack.newDefaultUnpacker(payload.getBytes(StandardCharsets.ISO_8859_1));
+            while (unpacker.hasNext()) {
+                int itemCount = unpacker.unpackArrayHeader();
+                HubMessageType messageType = HubMessageType.values()[unpacker.unpackInt() - 1];
+                
+                switch (messageType) {
+                    case INVOCATION:
+                        hubMessages.add(createInvocationMessage(unpacker, binder, itemCount));
+                        break;
+                    case STREAM_ITEM:
+                        hubMessages.add(createStreamItemMessage(unpacker, binder));
+                        break;
+                    case COMPLETION:
+                        hubMessages.add(createCompletionMessage(unpacker, binder));
+                        break;
+                    case STREAM_INVOCATION:
+                        hubMessages.add(createStreamInvocationMessage(unpacker, binder, itemCount));
+                        break;
+                    case CANCEL_INVOCATION:
+                        hubMessages.add(createCancelInvocationMessage(unpacker));
+                        break;
+                    case PING:
+                        hubMessages.add(PingMessage.getInstance());
+                        break;
+                    case CLOSE:
+                        hubMessages.add(createCloseMessage(unpacker, itemCount));
+                        break;
+                    default:
+                        break;
+                }
+            }
+        } catch (MessagePackException | IOException ex) {
+            throw new RuntimeException("Error reading MessagePack data.", ex);
+        }
+
+        return hubMessages.toArray(new HubMessage[hubMessages.size()]);
```

Comment:
Wonder if we should change the method to return a List so we don't need this extra allocate and copy

Reply:
Done https://github.com/dotnet/aspnetcore/pull/23532/commits/6a569b9d0a224c98a49c5658f929f0b64d77364c